slick/src/main/scala/slick/util/ManagedArrayBlockingQueue.scala

```diff
@@ -1,87 +1,68 @@
+import java.util
+
+import slick.util.AsyncExecutor.{LowPriority, MediumPriority, HighPriority, Priority}
 
 /** A simplified copy of `java.util.concurrent.ArrayBlockingQueue` with additional logic for
   * temporarily rejecting elements based on the current size. All features of the original
   * ArrayBlockingQueue have been ported, except the mutation methods of the iterator. See
   * `java.util.concurrent.ArrayBlockingQueue` for documentation. */
-abstract class ManagedArrayBlockingQueue[E >: Null <: AnyRef](capacity: Int, fair: Boolean = false) extends AbstractQueue[E] with BlockingQueue[E] { self =>
+abstract class ManagedArrayBlockingQueue[E >: Null <: AnyRef](capacity: Int, fair: Boolean = false)
+  extends util.AbstractQueue[E]
+  with BlockingQueue[E]
+  with Pausable { self =>
 
-  /** Determine if the item should be accepted at the current time. */
-  protected[this] def accept(item: E, size: Int): Boolean
+  protected[this] def priority(item: E): Priority
 
-  private[this] val items = new Array[AnyRef](capacity)
   private[this] val lock = new ReentrantLock(fair)
   private[this] val notEmpty = lock.newCondition
   private[this] val notFull = lock.newCondition
-  private[this] var takeIndex, putIndex, count = 0
 
   private[this] def checkNotNull(v: AnyRef): Unit = if (v == null) throw new NullPointerException
 
-  private[this] def inc(i: Int): Int = if(i+1 == items.length) 0 else i+1
-
-  private[this] def dec(i: Int): Int = (if(i == 0) items.length else i) - 1
+  private[this] val itemQueue = new InternalArrayQueue[E](2*capacity)
+  private[this] val highPrioItemQueue = new InternalArrayQueue[E](capacity)
 
-  private[this] def itemAt(i: Int): E = items(i).asInstanceOf[E]
+  private[this] def counts = (if (paused) 0 else  itemQueue.count) + highPrioItemQueue.count
 
-  private[this] def insert(x: E) {
-    items(putIndex) = x
-    putIndex = inc(putIndex)
-    count += 1
-    notEmpty.signal
-  }
+  private[this] var paused = false
 
-  private[this] def extract: E = {
-    val items = this.items
-    val x: E = items(takeIndex).asInstanceOf[E]
-    items(takeIndex) = null
-    takeIndex = inc(takeIndex)
-    count -= 1
-    notFull.signal
-    return x
+  /**
+    * Pausing causes only HighPriority items to be processed
+    */
+  def pause() = {
+    locked { paused = true }
   }
 
-  private[this] def removeAt(_i: Int) {
-    var i = _i
-    val items = this.items
-    if (i == takeIndex) {
-      items(takeIndex) = null
-      takeIndex = inc(takeIndex)
-    }
-    else {
-      var cond = true
-      while (cond) {
-        val nexti: Int = inc(i)
-        if (nexti != putIndex) {
-          items(i) = items(nexti)
-          i = nexti
-        }
-        else {
-          items(i) = null
-          putIndex = i
-          cond = false
-        }
-      }
-    }
-    count -= 1
-    notFull.signal
+  /**
+    * Resuming switches back to processing all items
+    */
+  def resume() = {
+    locked { paused = false }
```

Comment:
As explained above, pausing does not completely pause the threadpool, but only allow high-priority items to be taken. If there are high-priority items then take will not block and return a high-priority item, which upon releasing the connection will call resume. If there are no more high-priority items left it will wait for notEmpty.

Maybe resume() should besides setting pause to false also call notEmpty.signalAll if there are any lower priority items left?